shell/platform/android/io/flutter/embedding/engine/FlutterJNI.java

```diff
@@ -939,6 +942,112 @@ public void setLocalizationPlugin(@Nullable LocalizationPlugin localizationPlugi
 
   // ----- End Localization Support ----
 
+  // ----- Start Dynamic Features Support ----
+
+  /** Sets the dynamic feature manager that is used to download and install split features. */
+  @UiThread
+  public void setDynamicFeatureManager(@NonNull DynamicFeatureManager dynamicFeatureManager) {
+    ensureRunningOnMainThread();
+    this.dynamicFeatureManager = dynamicFeatureManager;
+  }
+
+  private Context dynamicFeatureContext;
+
+  @UiThread
+  public void setDynamicFeatureContext(@NonNull Context context) {
+    ensureRunningOnMainThread();
+    this.dynamicFeatureContext = context;
+  }
+
+  @SuppressWarnings("unused")
+  @UiThread
+  public void downloadDynamicFeature(int loadingUnitId) {
+    String loadingUnitIdResName =
+        dynamicFeatureContext
+            .getResources()
+            .getString(
+                dynamicFeatureContext
+                    .getResources()
+                    .getIdentifier(
+                        "loadingUnit" + loadingUnitId,
+                        "string",
+                        dynamicFeatureContext.getPackageName()));
+    downloadDynamicFeature(loadingUnitIdResName, loadingUnitId);
+  }
+
+  // Called by the engine upon invocation of dart loadLibrary() request
+  @SuppressWarnings("unused")
+  @UiThread
+  public void downloadDynamicFeature(String moduleName, int loadingUnitId) {
+    dynamicFeatureManager.downloadFeature(moduleName, loadingUnitId);
+  }
+
+  /**
+   * This should be called for every loading unit to be loaded into the dart isolate.
+   *
+   * <p>abi, libName, and apkPaths are used together to search the installed apks for the desired
+   * .so library. If not found, soPath may be provided as a fallback if a pre-extracted .so exists,
+   * especially on older devices with libs compressed in the apk.
+   *
+   * <p>Successful loading of the dart library also completes the loadLibrary() future that
+   * triggered the install/load process.
+   */
+  @UiThread
+  public void loadDartLibrary(
+      int loadingUnitId,
+      @NonNull String libName,
+      @NonNull String[] apkPaths,
+      @NonNull String abi,
+      @NonNull String soPath) {
+    ensureRunningOnMainThread();
+    ensureAttachedToNative();
+    nativeLoadDartLibrary(nativePlatformViewId, loadingUnitId, libName, apkPaths, abi, soPath);
+  }
+
+  private native void nativeLoadDartLibrary(
+      long nativePlatformViewId,
+      int loadingUnitId,
+      @NonNull String libName,
+      @NonNull String[] apkPaths,
+      @NonNull String abi,
+      @NonNull String soPath);
+
+  /**
+   * Specifies a new AssetManager that has access to the dynamic feature's assets in addition to the
+   * base module's assets.
+   *
+   * <p>assetBundlePath is the subdirectory that the flutter assets are stored in. The typical value
+   * is `flutter_assets`.
+   */
+  @UiThread
+  public void updateAssetManager(
```

Comment:
I may have to add additional capability to the engine's asset manager to full support hot reloading and whatnot. I'll leave this work to a follow-up PR.

Reply:
I might not have fully understood the reply since the term "asset manager" is very overloaded :) 

My comment was just generally on supporting the spirit of https://github.com/flutter/engine/blob/master/shell/platform/android/io/flutter/embedding/engine/FlutterJNI.java#L47.

Reply:
I think it would still be ideal for the runBundleAndSnapshotFromLibrary's AssetManager to just be a router. Then the re-routing can just happen purely on the Java side without having to pass another thing into JNI. In the spirit of having as few JNI methods as possible.